File: pset2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

bool is_end_of_sentence(int);
void print_grade(int);

int main(void) {
    string s = get_string("Text: ");
    int len = strlen(s);
    int letters = 0;
    int words = 0;
    int sentences = 0;
    bool inside_word = false;
    int chr;

    for (int i = 0; i < len; i++) {
        chr = s[i];

        if (isalpha(chr)) {
            if (!inside_word) inside_word = true;
            letters++;
        }

        if ((inside_word && chr == ' ') || is_end_of_sentence(chr)) {
            inside_word = false;
            words++;
        }

        if (is_end_of_sentence(chr)) sentences++;
    }

    float L = (float) letters / (float) words * 100;
    float S = (float) sentences / (float) words * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;
    int grade = round(index);

    print_grade(grade);
}

bool is_end_of_sentence(int chr) {
    return chr == '.' || chr == '!' || chr == '?';
}

void print_grade(int grade) {
    if (grade >= 16) {
        printf("Grade 16+\n");
    } else if (grade < 1) {
        printf("Before Grade 1\n");
    } else {
        printf("Grade %i\n", grade);
    }
}

File: pset6/readability.py
import cs50
import re


def main():
    text = cs50.get_string('Text: ')
    index = get_readability_index(text)
    print_grade(index)


def get_readability_index(text):
    letters = len(re.findall(r'[a-zA-Z]', text))
    words = len(re.findall(r'\S+', text))
    sentences = len(re.findall(r'[.?!]', text))

    L = letters / words * 100
    S = sentences / words * 100
    index = 0.0588 * L - 0.296 * S - 15.8

    return index


def print_grade(index):
    grade = round(index)

    if grade >= 16:
        print('Grade 16+')
    elif grade < 1:
        print('Before Grade 1')
    else:
        print(f'Grade {grade}')


if __name__ == '__main__':
    main()
